File: salescompass/backend/app/schemas/analysis.py
from datetime import datetime
from typing import Any

from pydantic import BaseModel, ConfigDict, Field, model_validator

from app.schemas.company import CompanyCreate, CompanyMode, CompanyRead


class SegmentScore(BaseModel):
    name: str
    score: int = Field(ge=0, le=100)
    fit: int = Field(ge=0, le=100)
    urgency: int = Field(ge=0, le=100)
    reachability: int = Field(ge=0, le=100)
    deal_quality: int = Field(ge=0, le=100)
    evidence: list[str]


class OutreachVariation(BaseModel):
    title: str
    channel: str
    message: str


class AnalysisResult(BaseModel):
    diagnosis: str
    recommended_icp: str
    confidence: int = Field(ge=0, le=100)
    market_scores: list[SegmentScore]
    disqualifiers: list[str]
    external_benchmarks: list[str]
    action_plan: list[str]
    outreach: list[OutreachVariation]
    human_checkpoint: str
    assumptions: list[str]


class AnalysisCreate(BaseModel):
    company: CompanyCreate | None = None
    company_id: int | None = None
    mode: CompanyMode | None = None
    input: dict[str, Any] | None = None

    @model_validator(mode="after")
    def require_analysis_input(self) -> "AnalysisCreate":
        if self.company is None and self.company_id is None and self.input is None:
            raise ValueError("Provide either company, company_id, or input")
        return self


class AnalysisRefineRequest(BaseModel):
    notes: str = Field(min_length=1, max_length=4000)
    input: dict[str, Any] | None = None


class ActionPlanResponse(BaseModel):
    run_id: int
    action_plan: dict[str, Any] | list[Any]


class ICPRunRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    run_id: int
    user_id: int
    company_id: int
    status: str
    mode: str
    input_snapshot: dict[str, Any]
    agent_output: AnalysisResult
    baseline_output: dict[str, Any]
    action_plan: dict[str, Any] | list[Any] | None = None
    refinement_notes: str | None = None
    error_message: str | None = None
    result: AnalysisResult
    model_name: str
    created_at: datetime
    updated_at: datetime
    completed_at: datetime | None
    company: CompanyRead | None = None

File: salescompass/frontend/src/types/analysis.ts
import type { Company } from "./company";

export type CompanyMode = "history" | "no_history";

export type ConfidenceLevel = "low" | "medium" | "high";

export type AnalysisStatus = "pending" | "completed" | "failed";

export type ReviewStatus = "needs_review" | "approved";

// Raw backend API types. Keep these aligned to JSON from /analyses.
export type SegmentScoresApi = {
  size?: number;
  access?: number;
  ticket?: number;
  cycle?: number;
  competition?: number;
};

export type MarketSegmentApi = {
  name?: string;
  scores?: SegmentScoresApi;
  total?: number | null;
  rationale?: string;
  [key: string]: unknown;
};

export type ICPOutputApi = {
  profile?: string;
  company_size?: string;
  target_industry?: string;
  region?: string;
  decision_maker?: string;
  main_pain?: string;
  rationale?: string;
  confidence?: ConfidenceLevel | string;
  confidence_basis?: string;
  [key: string]: unknown;
};

export type OutreachOutputApi = {
  channel?: string;
  trigger?: string;
  first_contact?: string;
  message_tone?: string;
  sample_message?: string;
  confidence?: ConfidenceLevel | string;
  confidence_basis?: string;
  [key: string]: unknown;
};

export type BenchmarkApi = unknown;

export type AgentOutputApi = {
  diagnosis?: string;
  external_benchmarks?: BenchmarkApi[];
  markets?: MarketSegmentApi[];
  icp?: ICPOutputApi;
  approach?: OutreachOutputApi;
  hypotheses_to_validate?: string[];
  questions_for_human?: string[];
};

export type BaselineOutputApi = {
  segment?: string;
  icp?: string;
  rationale?: string;
  confidence?: ConfidenceLevel | string | number;
  outreach?: {
    channel?: string;
    message?: string;
  };
  [key: string]: unknown;
};

export type ActionPlanApi = Record<string, unknown> | unknown[];

export type AnalysisRunApi = {
  id?: number;
  run_id?: number;
  user_id?: number;
  company_id?: number;
  status?: AnalysisStatus | string;
  mode?: CompanyMode | string;
  input_snapshot?: Record<string, unknown>;
  agent_output?: AgentOutputApi;
  baseline_output?: BaselineOutputApi;
  action_plan?: ActionPlanApi | null;
  review_status?: ReviewStatus | string;
  refinement_notes?: string | null;
  error_message?: string | null;
  model_name?: string;
  created_at?: string;
  updated_at?: string;
  completed_at?: string | null;
  company?: Company | null;
};

// Frontend view-model types. These are normalized for rendering only.
export type SegmentScoresView = {
  size: number;
  access: number;
  ticket: number;
  cycle: number;
  competition: number;
};

export type MarketSegmentView = {
  name: string;
  total: number;
  rationale: string;
  scores: SegmentScoresView;
};

export type RecommendedICPView = {
  profile: string;
  industry: string;
  companySize: string;
  region: string;
  decisionMaker: string;
  painPoint: string;
  rationale: string;
  confidence: ConfidenceLevel;
  confidenceBasis: string;
};

export type OutreachView = {
  channel: string;
  trigger: string;
  firstContact: string;
  tone: string;
  sampleMessage: string;
  confidence: ConfidenceLevel;
  confidenceBasis: string;
};

export type BenchmarkView = {
  stat: string;
  source: string;
};

export type BaselineView = {
  segment: string;
  icp: string;
  rationale: string;
  confidence: ConfidenceLevel;
  outreachChannel: string;
  outreachMessage: string;
};

export type EvidenceItemView = {
  label: string;
  value: string;
};

export type DowngradedSegmentView = {
  name: string;
  score: number;
  rationale: string;
};

export type ModeEvidenceView = {
  salesHistory: EvidenceItemView[];
  wonLostPatterns: EvidenceItemView[];
  conversionEvidence: EvidenceItemView[];
  downgradedSegments: DowngradedSegmentView[];
  marketAssumptions: EvidenceItemView[];
  demoMarketContext: EvidenceItemView[];
};

export type ActionPlanStepView = {
  title: string;
  owner?: string;
  timeframe?: string;
  successMetric?: string;
};

export type ActionMessageVariationView = {
  title: string;
  channel: string;
  message: string;
};

export type ActionPlanView = {
  summary?: string;
  nextSteps: ActionPlanStepView[];
  messageVariations: ActionMessageVariationView[];
  metricsToTrack: string[];
};

export type AnalysisViewModel = {
  runId: number;
  companyId?: number;
  companyName?: string;
  status: AnalysisStatus;
  mode: CompanyMode;
  createdAt?: string;

  diagnosis: string;
  benchmarks: BenchmarkView[];
  markets: MarketSegmentView[];
  recommendedICP: RecommendedICPView;
  outreach: OutreachView;
  hypothesesToValidate: string[];
  questionsForHuman: string[];
  baseline: BaselineView;
  modeEvidence: ModeEvidenceView;
  actionPlan?: ActionPlanView;
  reviewStatus: ReviewStatus;
  errorMessage?: string | null;
};
